### PR TITLE
ENH: remove fallback needed only with now unsupported Meson versions

### PR DESCRIPTION
**Pull request by dnicolodi**
I think we can safely drop this now. The CI should agree.

### CODE REVIEW

File: mesonpy/__init__.py
Comment (rgommers) [inline mesonpy/__init__.py:407]: Under what circumstances is returning `None` the correct answer? If it's never the right answer, then should this raise an exception?

**Comment by rgommers** [comment]
Thanks @dnicolodi. I'm a little nervous about this one, because we've had so many problems with this logic before. I think all Meson issues are fixed indeed. However, it's hard for me to verify that we definitely won't need this fallback code anymore. IIRC this was a problematic case that we may not have a test for:

```meson
custom_target('file_with_random_extension',
  input : ['codegen.c.in'],
  output : ['codegen.c'],
  command : [custom_script, '@INPUT@', '@OUTPUT@'],
  install : true,  # still needed because of meson#7372 - can't pass generated targets to py.install_sources
  install_dir : py.get_install_dir() / 'pkgname' / 'submod'  # this may work now, not sure if that covers everything
```

Or are we sure that by design there is no way to hit that heuristics code path anymore?

File: mesonpy/__init__.py
Comment (dnicolodi) [reply mesonpy/__init__.py:407]: Where this function is called a warning is emitted if the return value is `None`. I think we can turn this into an error, but I didn't want to modify the current code too much.

**Comment by dnicolodi** [comment]
I just verified that the case above works correctly.

**Comment by rgommers** [comment]
I tested this PR with SciPy, and all works as expected for a regular wheel build on Linux. So that looks good. I'll leave further review to @FFY00, because he's more familiar with this code.

**Comment by FFY00** [comment]
Yeah, I am a bit uneasy about this. IMO we should wait until after the editable installs release to merge this, unless there's any dependence on this[^1].

[^1]: The `Traversable` approach would have a dependence on this, or at least, would greatly benefit from this because it needs to map the files. But I'd prefer to also wait to merge that change.

**Comment by FFY00** [comment]
Even though newer Meson versions have it fixed, several users will still be using older versions, shipped by their distros. I'd really prefer to do this change after we release editable installs, so that we don't have to deal with breakage coming from several places.

> If we want to err on the safe side, we can check that all installation paths start with a placeholder and error out if they do not.

Yes, we should do that.

**Comment by FFY00** [comment]
Please note that this change has a possibility of breaking normal installs. I just want to wait a little bit.

**Comment by dnicolodi** [comment]
No, they cannot use older Meson shipped by their distros, we depend on Meson >= 0.63.3: https://github.com/mesonbuild/meson-python/blob/c9438b0c94fac9672173714c6de2d88a99c4e2e4/pyproject.toml#L31 We test this version an all sebsequent minor versions in the CI.

We already emit a warning if we cannot map the file. We can turn it into an error if the path does not start with a placeholder.

**Comment by dnicolodi** [comment]
I'm not in an hurry to merge this, but for what should we wait exactly? To say that we need to wait without specifying what we need to wait for seem like postponing indefinitely.

What we should have done, was to add comments to this code specifying in which cases it was required, and ideally we should have added corresponding tests to the test suite. Now we would not be debating whether this work-around is still needed or not. But "del senno di poi son piene le fosse".

**Comment by FFY00** [comment]
> No, they cannot use older Meson shipped by their distros, we depend on Meson >= 0.63.3:

I forgot about that.

File: mesonpy/__init__.py
Comment (FFY00) [reply mesonpy/__init__.py:407]: The reason for that was that we had two functions to try to map the files (via the direct scheme mapping, and heuristics). Now, we can move the warning here.

---

What was the reason for changing the mapping logic here? I mean, I am not gonna block it, but next time please keep keep code improvements separate from logical changes.

File: mesonpy/__init__.py
Comment (FFY00) [inline mesonpy/__init__.py:406]:
```suggestion
                return scheme, pathlib.Path(*parts[1:])
```

Also, please keep the docstring.

**Review by FFY00 — CHANGES_REQUESTED** [review]

File: mesonpy/__init__.py
Comment (dnicolodi) [reply mesonpy/__init__.py:406]: Oh, neat. I'm actually quite sure I tried this syntax and that it didn't work. I must have forgot the `*` or something.

File: mesonpy/__init__.py
Comment (dnicolodi) [reply mesonpy/__init__.py:407]: I had to read the previous version three times to make sure of what it does. Having nested iterator comprehensions in a `for` loop is just too much for me. So I decided to write in a way that I find more immediate to read.

it is a separate commit. Do you mean I should open a separate PR next time?

Comment (dnicolodi) [reply mesonpy/__init__.py:407]: Cool. We broke GitHub interface with this comments. They do not show up in the main PR page...

File: mesonpy/__init__.py
Comment (FFY00) [reply mesonpy/__init__.py:406]: Yeah, the constructor can't take a sequence as the first arguments, so you need to unpack it.

File: mesonpy/__init__.py
Comment (FFY00) [reply mesonpy/__init__.py:407]: No, separate commit is fine. I thought you were planning to squash-merge because the commit message was so small :sweat_smile:. Next time, use something like `simplify _map_from_scheme_map`.

**Comment by FFY00** [comment]
> Cool. We broke GitHub interface with this comments. They do not show up in the main PR page...

@dnicolodi did you delete this? Because I actually do not see your comment at all :rofl:. I just got the email notification.

**Comment by dnicolodi** [comment]
> @dnicolodi did you delete this? Because I actually do not see your comment at all 🤣. I just got the email notification.

Yes, I did. The comments are actually there. Just in a weird order. Sometimes the GitHub interface confuses me.

**Comment by rgommers** [comment]
> > No, they cannot use older Meson shipped by their distros, we depend on Meson >= 0.63.3:
> 
> I forgot about that.

We're missing `meson_version: '>= 0.63.3' in the `project()` call in our own `meson.build` file, so this is right now not quite correct for non-isolated builds. Can we please add that to ensure we don't use an older Meson version? Just add a comment that the minimum version is implemented in both `pyproject.toml` and `meson.build` - that's unavoidable, and updating in two places is a very small effort.

**Comment by rgommers** [comment]
(sorry, wrong button)

**Comment by dnicolodi** [comment]
> We're missing `meson_version: '>= 0.63.3'` in the `project()` call in our own `meson.build` file, so this is right now not quite correct for non-isolated builds. Can we please add that to ensure we don't use an older Meson version? Just add a comment that the minimum version is defined in both `pyproject.toml` and `meson.build` - that's unavoidable, and updating in two places is a very small effort.

Adding something to the meson-python's `meson.build` enforces the minimum version only when building meson-python.

However, this is a good point. It is not impossible to conceive a way to use the system meson while installing meson-python in a virtual environment. I think maybe we should add a run-time check for the meson version. I think it would cost very little. Going down this line we could even move meson to a build dependency of the project being built obtained via `get_requires_for_build_wheel()` as we did for ninja, only if the meson executable is not found or is too old.

**Comment by rgommers** [comment]
> However, this is a good point. It is not impossible to conceive a way to use the system meson while installing meson-python in a virtual environment. I think maybe we should add a run-time check for the meson version.

You are right, good point. +1 for a check.

**Comment by eli-schwartz** [comment]
> Just add a comment that the minimum version is defined in both `pyproject.toml` and `meson.build` - that's unavoidable, and updating in two places is a very small effort.

> Going down this line we could even move meson to a build dependency of the project being built obtained via `get_requires_for_build_wheel()` as we did for ninja, only if the meson executable is not found or is too old.


It is theoretically possible to parse `meson.build` for the `meson_version` once you have some version of Meson installed, to fine-tune the required version number. Not sure how helpful that is in practice?

```python
import json
import subprocess

stdout = subprocess.check_output(['meson', 'introspect', '--ast', 'meson.build'])
j = json.loads(stdout)

for kw in j['lines'][0]['args']['kwargs']:
    if kw['key']['value'] == 'meson_version':
        meson_version = kw['val']['value']
        break
```

**Comment by dnicolodi** [comment]
We don't install the meson-python's `meson.build` thus extracting the minimum required meson version from there at runtime is not that useful. Unfortunately, I think that having the minimum version both in `pyproject.toml` and somewhere in the code is the only way to do the check at run time (with reasonable complexity).

**Comment by dnicolodi** [comment]
Is there a way to make the output of pre-commit-check more informative? pre-commit-check is failing with this error
```
ruff.....................................................................Failed
- hook id: ruff
- files were modified by this hook

Found 1 error (1 fixed, 0 remaining).
```
which tells me exactly nothing about what is wrong.

**Comment by dnicolodi** [comment]
Why can't this be merged as is? It just removes a fallback for a bug in version of Meson that are not supported by meson-python anyway. I don't think that waiting will make anyone more confident that the bugs have actually been resolved. Do we need more tests? What configurations not covered by the tests already in place do we need to test? I would like to merge this and move on to fix #317.

File: mesonpy/__init__.py
Comment (FFY00) [inline mesonpy/__init__.py:1024]:
```suggestion
        return tuple(map(int, string.split('.')))
```

File: mesonpy/__init__.py
Comment (FFY00) [inline mesonpy/__init__.py:1126]: This is better placed in `Project`, like ninja.

File: mesonpy/_introspection.py
Comment (FFY00) [inline mesonpy/_introspection.py:71]: You can remove the rest of the Debian introspection code too.

**Review by FFY00 — CHANGES_REQUESTED** [review]

File: mesonpy/__init__.py
Comment (dnicolodi) [reply mesonpy/__init__.py:165]: > Please keep these functions in `_WheelBuilder`.

Why? They don't need access to any member of `_WheelBuilder`.

> IMO it'd be cleaner to move the warnings to `_map_meson_destination`.

For doing that, either the src path needs to be passed to `_map_meson_destination` just for the purpose of maybe emitting the warning this information needs to be removed from the warning.

Comment (FFY00) [reply mesonpy/__init__.py]: > Why? They don't need access to any member of `_WheelBuilder`.

Because its whole purpose is to support code in `_WheelBuilder`, and this way it also can access internal state, for example the source paths you referred in the next comment.

> For doing that, either the src path needs to be passed to `_map_meson_destination` just for the purpose of maybe emitting the warning this information needs to be removed from the warning.

Or you can have it in `_WheelBuilder` and be able to access it directly :sweat_smile:

Comment (dnicolodi) [reply mesonpy/__init__.py]: Please have a look at how the code is written. Having it a method of `_WheelBuilder` does not help in any way.

**Comment by rgommers** [comment]
I did some more testing with the merged version of this in SciPy's CI. We have a specific job for Debian there which used to be an issue. All good now it looks like. Glad to see this code gone!

**Pull request merged**
